File: tazapp/src/main/java/de/thecode/android/tazreader/dialog/HelpDialog.java
package de.thecode.android.tazreader.dialog;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebChromeClient;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import de.mateware.dialog.DialogCustomView;
import de.thecode.android.tazreader.R;
import de.thecode.android.tazreader.data.DownloadState;
import de.thecode.android.tazreader.data.Paper;
import de.thecode.android.tazreader.data.PaperRepository;
import de.thecode.android.tazreader.data.Resource;
import de.thecode.android.tazreader.data.ResourceRepository;
import de.thecode.android.tazreader.utils.AsyncTaskListener;
import de.thecode.android.tazreader.utils.StorageManager;

import java.io.File;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.List;

import androidx.annotation.Nullable;
import androidx.annotation.StringDef;
import timber.log.Timber;

/**
 * Created by mate on 25.07.2017.
 */

public class HelpDialog extends DialogCustomView {

    @Retention(RetentionPolicy.SOURCE)
    @StringDef({HELP_INTRO, HELP_LIBRARY, HELP_ARTICLE, HELP_PAGE, HELP_PRIVACY})
    public @interface HelpPage {
    }


    public static final String HELP_INTRO   = "intro.html";
    public static final String HELP_LIBRARY = "ausgaben.html";
    public static final String HELP_PAGE    = "seiten.html";
    public static final String HELP_ARTICLE = "artikel.html";
    public static final String HELP_PRIVACY = "datenschutz.html";

    private static final String[] HELP_RESOURCE_SUBDIRS = {"res/android-help", "res/ios-help"};

    private static final String ARG_HELPPAGE   = "helpPage";
    private static final String ARG_CURRENTURL = "currentUrl";
    private static final String ARG_BASEURL    = "baseUrl";

    private String  baseUrlPath;
    private WebView webView;
    private String  currentUrl;

    private PaperRepository    paperRepository;
    private ResourceRepository resourceRepository;
    private StorageManager     storageManager;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (savedInstanceState != null) {
            currentUrl = savedInstanceState.getString(ARG_CURRENTURL);
            baseUrlPath = savedInstanceState.getString(ARG_BASEURL);
        }
    }

    @SuppressLint("SetJavaScriptEnabled")
    @Override
    public View getView(final LayoutInflater inflater, ViewGroup parent) {

        View view = inflater.inflate(R.layout.dialog_help, parent, false);
        webView = view.findViewById(R.id.help_web_view);
        webView.getSettings().setJavaScriptEnabled(true);
        webView.setWebViewClient(new WebViewClient() {

            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                Timber.i("%s", url);
                if (url != null && url.startsWith(baseUrlPath)) {
                    view.loadUrl(url);
                } else {
                    Intent i = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
                    view.getContext()
                        .startActivity(i);
                }
                return true;
            }
        });
        webView.setWebChromeClient(new WebChromeClient());

        if (TextUtils.isEmpty(currentUrl)) {
            paperRepository = PaperRepository.getInstance(view.getContext());
            resourceRepository = ResourceRepository.getInstance(view.getContext());
            storageManager = StorageManager.getInstance(view.getContext());
            new AsyncTaskListener<Void, String>(aVoid -> {
                List<Paper> papers = paperRepository.getAllPapers();
                String result = "file:///android_asset/help/";
                for (Paper paper : papers) {
                    Resource latestResource = resourceRepository.getWithKey(paper.getResource());
                    if (latestResource != null && resourceRepository.getDownloadState(latestResource.getKey()) == DownloadState.READY) {
                        File latestResourceDir = storageManager.getResourceDirectory(latestResource.getKey());
                        File helpFileDir = null;
                        for (String helpFileSubdirPath : HELP_RESOURCE_SUBDIRS) {
                            helpFileDir = new File(latestResourceDir, helpFileSubdirPath);
                            if (helpFileDir.exists()) break;
                        }
                        if (helpFileDir != null) {
                            result = "file://" + helpFileDir.getAbsolutePath() + "/";
                            break;
                        }
                    }
                }
                return result;
            }, baseUrl -> {
                baseUrlPath = baseUrl;
                String helpPage = getArguments().getString(ARG_HELPPAGE);
                currentUrl = baseUrlPath + helpPage;
                loadUrl();
            }).execute();
        } else {
            loadUrl();
        }

        return view;
    }

    private void loadUrl() {
        Timber.d("loading url %s", currentUrl);
        webView.loadUrl(currentUrl);
    }

    @Override
    public Bundle onSaveInstanceState(Bundle outState) {
        if (webView != null) outState.putString(ARG_CURRENTURL, webView.getUrl());
        outState.putString(ARG_BASEURL, baseUrlPath);
        return super.onSaveInstanceState(outState);
    }

    public static class Builder extends AbstractBuilder<Builder, HelpDialog> {

        Bundle bundle = new Bundle();

        public Builder() {
            super(HelpDialog.class);
        }

        public Builder setHelpPage(@HelpPage String helpPage) {
            bundle.putString(ARG_HELPPAGE, helpPage);
            return this;
        }

        @Override
        public void preBuild() {
            super.preBuild();
            addBundle(bundle);
        }
    }

}

File: tazapp/src/main/java/de/thecode/android/tazreader/dialognew/HelpDialog.kt
package de.thecode.android.tazreader.dialognew

import android.app.Dialog
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.view.View
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.*
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.customview.customView
import com.afollestad.materialdialogs.customview.getCustomView
import com.github.ajalt.timberkt.d
import de.thecode.android.tazreader.R
import de.thecode.android.tazreader.data.DownloadState
import de.thecode.android.tazreader.paperRepository
import de.thecode.android.tazreader.resourceRepository
import de.thecode.android.tazreader.storageManager
import org.jetbrains.anko.doAsync
import java.io.File

class HelpDialog : DialogFragment() {

    companion object {
        const val TAG = "helpDialog"
        const val ARG_PAGE = "helpPage"
        fun newInstance(page: HelpPage): HelpDialog {
            val bundle = Bundle()
            bundle.putSerializable(ARG_PAGE, page)
            val fragment = HelpDialog()
            fragment.arguments = bundle
            return fragment
        }
    }

    var webView: WebView? = null

    private val viewModel: HelpDialogViewModel by lazy {
        ViewModelProviders.of(this, HelpDialogViewModelFactory(arguments!!.getSerializable(ARG_PAGE) as HelpPage))
                .get(HelpDialogViewModel::class.java)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        d { "XXXX onCreateDialog" }
        context?.let { dialogContext ->
            val dialog = MaterialDialog(dialogContext).customView(viewRes = R.layout.dialog_help,noVerticalPadding = true,dialogWrapContent = true)
            dialog.maxWidth(R.dimen.help_max_width)
            dialog.positiveButton()
            if (viewModel.page == HelpPage.INTRO){
                dialog.neutralButton(R.string.drawer_account) {
                    activity?.let {
                        if (it is HelpIntroActivity){
                            it.onAccountDataClick()
                        }
                    }
                }
            }
            webView = dialog.getCustomView()
                    .findViewById(R.id.help_web_view)
            webView?.settings!!.javaScriptEnabled = true
            webView?.webViewClient = object : WebViewClient() {
                override fun shouldOverrideUrlLoading(view: WebView?, url: String?): Boolean {
                    url?.let {
                        if (url.startsWith(viewModel.baseUrl)) {
                            view?.loadUrl(url)
                        } else {
                            context?.startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(url)))
                        }
                    }
                    return true
                }

                override fun onPageFinished(view: WebView?, url: String?) {
                    d { "XXX onPageFinished $url" }
                    viewModel.currentUrlLiveData.value = url
                    super.onPageFinished(view, url)
                }
            }
            viewModel.currentUrlLiveData.observe(this, Observer { urlString ->
                d { "XXX ${webView!!.url}" }
                webView?.let {
                    if (it.url != urlString) {
                        webView?.loadUrl(urlString)
                    }
                }
            })
            return dialog
        }

        return super.onCreateDialog(savedInstanceState)

    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        d { "XXXX onViewCreated" }
        super.onViewCreated(view, savedInstanceState)
    }
}

enum class HelpPage(val pageName: String) {
    INTRO("intro.html"),
    LIBRARY("ausgaben.html"),
    PAGE("help.html#seiten"),
    ARTICLE("artikel.html"),
    PRIVACY("datenschutz.html");
}

class HelpDialogViewModel(val page: HelpPage) : ViewModel() {

    var baseUrl = "file:///android_asset/help/"
    val currentUrlLiveData = MutableLiveData<String>()

    init {
        //currentUrlLiveData.value = "file:///android_asset/help/${page.pageName}"
        doAsync {
            run finder@{
                paperRepository.allPapers.forEach { paper ->
                    val latestResource = resourceRepository.getResourceForPaper(paper)
                    latestResource?.let { resource ->
                        d { "XXX $latestResource" }
                        if (resource.downloadState == DownloadState.READY) {
                            val latestResourceDir = storageManager.getResourceDirectory(resource)
                            d { "XXX $latestResourceDir" }
                            arrayOf("res/android-help", "res/ios-help").map {
                                File(latestResourceDir, it)
                            }
                                    .find {
                                        it.exists()
                                    }
                                    ?.let {
                                        baseUrl = "file://${it.absolutePath}/"
                                        return@finder
                                    }

                        }
                    }
                }
            }
            currentUrlLiveData.postValue(baseUrl + page.pageName)
        }
    }
}

@Suppress("UNCHECKED_CAST")
class HelpDialogViewModelFactory(val page: HelpPage) : ViewModelProvider.Factory {
    override fun <T : ViewModel?> create(modelClass: Class<T>): T {
        return HelpDialogViewModel(page) as T
    }
}

interface HelpIntroActivity {
    fun onAccountDataClick()
}
